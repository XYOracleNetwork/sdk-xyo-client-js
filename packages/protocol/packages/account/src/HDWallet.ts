import { generateMnemonic } from '@scure/bip39'
// eslint-disable-next-line import/no-internal-modules
import { wordlist as englishWordlist } from '@scure/bip39/wordlists/english'
import { toUint8Array } from '@xylabs/arraybuffer'
import { assertEx } from '@xylabs/assert'
import { Address, Hex, hexFromHexString } from '@xylabs/hex'
import { globallyUnique } from '@xylabs/object'
import { staticImplements } from '@xylabs/static-implements'
import {
  AccountConfig,
  isMnemonicInitializationConfig,
  isPhraseInitializationConfig,
  isPrivateKeyInitializationConfig,
} from '@xyo-network/account-model'
import type { PrivateKeyInstance } from '@xyo-network/key-model'
import { WalletInstance, WalletStatic } from '@xyo-network/wallet-model'
import { defaultPath, HDNodeWallet, Mnemonic } from 'ethers'

import { Account } from './Account.js'
import { PrivateKey } from './Key/index.js'

@staticImplements<WalletStatic>()
export class HDWallet extends Account implements WalletInstance {
  static override readonly uniqueName = globallyUnique('HDWallet', HDWallet, 'xyo')
  protected static override _addressMap: Record<Address, WeakRef<HDWallet>> = {}

  constructor(
    key: unknown,
    protected readonly node: HDNodeWallet,
    privateKey: PrivateKeyInstance,
  ) {
    super(Account._protectedConstructorKey, privateKey)
  }

  override get address(): Address {
    return hexFromHexString(this.node.address, { prefix: false })
  }

  override get addressBytes(): ArrayBuffer {
    return toUint8Array(this.address, undefined, 16)
  }

  get chainCode(): string {
    return this.node.chainCode
  }

  get depth(): number {
    return this.node.depth
  }

  get extendedKey(): string {
    return this.node.extendedKey
  }

  get fingerprint(): string {
    return this.node.fingerprint
  }

  get index(): number {
    return this.node.index
  }

  get mnemonic(): Mnemonic | null {
    return this.node.mnemonic
  }

  get parentFingerprint(): string {
    return this.node.parentFingerprint
  }

  get path(): string | null {
    return this.node.path
  }

  get privateKey(): Hex {
    return this.node.privateKey.toLowerCase() as Hex
  }

  get publicKey(): Hex {
    return this.node.publicKey.toLowerCase() as Hex
  }

  static override async create(opts?: AccountConfig): Promise<WalletInstance> {
    if (isPhraseInitializationConfig(opts)) {
      return await this.fromPhrase(opts.phrase)
    }
    if (isMnemonicInitializationConfig(opts)) {
      return await this.fromPhrase(opts.mnemonic, opts.path)
    }
    if (isPrivateKeyInitializationConfig(opts)) {
      throw new Error('Invalid initialization config. from privateKey not supported.  Use Account.fromPrivateKey instead.')
    }
    throw new Error('Invalid initialization config')
  }

  static async createFromNode(node: HDNodeWallet, previousHash?: string): Promise<WalletInstance> {
    return await this.createFromNodeInternal(node, previousHash)
  }

  static async fromExtendedKey(key: string): Promise<WalletInstance> {
    const node = HDNodeWallet.fromExtendedKey(key)
    /* TODO: Handle HDNodeVoidWallet */
    return await HDWallet.createFromNode(node as HDNodeWallet)
  }

  static async fromMnemonic(mnemonic: Mnemonic, path: string = defaultPath): Promise<WalletInstance> {
    const node = HDNodeWallet.fromMnemonic(mnemonic, path)
    const createdWallet = await this.createFromNodeInternal(node)
    return this.getCachedWalletOrCacheNewWallet(createdWallet)
  }

  static async fromPhrase(phrase: string, path: string = defaultPath): Promise<WalletInstance> {
    return await this.fromMnemonic(Mnemonic.fromPhrase(phrase), path)
  }

  static async fromSeed(seed: string | Uint8Array): Promise<WalletInstance> {
    const node = HDNodeWallet.fromSeed(seed)
    const createdWallet = await this.createFromNodeInternal(node)
    return this.getCachedWalletOrCacheNewWallet(createdWallet)
  }

  static generateMnemonic(wordlist: string[] = englishWordlist, strength: number = 256): string {
    return generateMnemonic(wordlist, strength)
  }

  static override async random(): Promise<WalletInstance> {
    return await this.fromMnemonic(Mnemonic.fromPhrase(HDWallet.generateMnemonic()))
  }

  protected static async createFromNodeInternal(node: HDNodeWallet, previousHash?: string): Promise<HDWallet> {
    const privateKey = toUint8Array(node.privateKey.replace('0x', ''))
    assertEx(!privateKey || privateKey?.length === 32, () => `Private key must be 32 bytes [${privateKey?.length}]`)
    const newWallet = await new HDWallet(Account._protectedConstructorKey, node, await PrivateKey.create(privateKey)).loadPreviousHash(previousHash)
    return HDWallet._addressMap[newWallet.address]?.deref() ?? newWallet
  }

  protected static getCachedWalletOrCacheNewWallet(createdWallet: HDWallet): HDWallet {
    const existingWallet = this._addressMap[createdWallet.address]?.deref()
    if (existingWallet) {
      return existingWallet
    }
    const ref = new WeakRef(createdWallet)
    this._addressMap[createdWallet.address] = ref
    return createdWallet
  }

  async derivePath(path: string): Promise<WalletInstance> {
    //if an absolute path, check if it matches the parent root and work with it
    if (path.startsWith('m/')) {
      const parentPath = this.path
      if (parentPath !== null && path.startsWith(parentPath)) {
        const childPath = path.slice(parentPath.length + 1)
        return await HDWallet.createFromNode(this.node.derivePath(childPath))
      }
      throw new Error(`Invalid absolute path ${path} for wallet with path ${parentPath}`)
    }
    return await HDWallet.createFromNode(this.node.derivePath(path))
  }

  neuter: () => HDWallet = () => {
    this.node.neuter()
    return this
  }
}
