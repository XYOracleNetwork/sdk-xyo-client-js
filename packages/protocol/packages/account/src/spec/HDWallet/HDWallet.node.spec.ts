import { HDWallet } from '../../HDWallet'
import { generateHDWalletTests } from './HDWallet.spec'

/**
 * @group nodejs
 */

describe.skip('Node Wallet Test', () => {
  generateHDWalletTests('HDWallet: Node', HDWallet)
})
