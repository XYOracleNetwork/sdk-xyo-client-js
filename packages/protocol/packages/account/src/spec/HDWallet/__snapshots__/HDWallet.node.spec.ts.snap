// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 1`] = `"012167a8cd67f0f20fbfc54212ac9bba982dbae3"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 2`] = `"0x012167A8cD67F0f20fBFC54212aC9BBa982Dbae3"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 3`] = `"6ab24a1bcccc3ff79e8f826b7cc5bf583b5e9c57"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 4`] = `"0x6AB24A1bccCc3FF79E8F826B7cc5BF583b5e9C57"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 5`] = `"5e0e54445fdd4e453fca8216b056226c5ef6a2f1"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 6`] = `"0x5e0E54445fDd4e453FCa8216b056226C5eF6a2F1"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 7`] = `"3654cafa0b49eef056adc7dce5f745b08f98498d"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 8`] = `"0x3654cafa0b49EEf056ADc7DCe5f745B08f98498D"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 9`] = `"f2a43815de64547920effab3112c6af940454fff"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 10`] = `"0xf2A43815dE64547920eFfAb3112C6AF940454fFf"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 11`] = `"e072bfa0adf20b46997aa346333683de09765ce6"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 12`] = `"0xe072bFA0ADF20b46997AA346333683de09765Ce6"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 13`] = `"09a9478624b423318d0b5c69b08896c9c961af4b"`;

exports[`Node Wallet Test HDWallet: Node constructor works repeatably & interoperably with Ethers 14`] = `"0x09A9478624b423318D0b5C69B08896C9C961af4B"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 1`] = `"012167a8cd67f0f20fbfc54212ac9bba982dbae3"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 2`] = `"012167a8cd67f0f20fbfc54212ac9bba982dbae3"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 3`] = `"6ab24a1bcccc3ff79e8f826b7cc5bf583b5e9c57"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 4`] = `"6ab24a1bcccc3ff79e8f826b7cc5bf583b5e9c57"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 5`] = `"5e0e54445fdd4e453fca8216b056226c5ef6a2f1"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 6`] = `"5e0e54445fdd4e453fca8216b056226c5ef6a2f1"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 7`] = `"3654cafa0b49eef056adc7dce5f745b08f98498d"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 8`] = `"3654cafa0b49eef056adc7dce5f745b08f98498d"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 9`] = `"f2a43815de64547920effab3112c6af940454fff"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 10`] = `"f2a43815de64547920effab3112c6af940454fff"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 11`] = `"e072bfa0adf20b46997aa346333683de09765ce6"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 12`] = `"e072bfa0adf20b46997aa346333683de09765ce6"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 13`] = `"09a9478624b423318d0b5c69b08896c9c961af4b"`;

exports[`Node Wallet Test HDWallet: Node derivePath works repeatably & interoperably from phrase & extended key 14`] = `"09a9478624b423318d0b5c69b08896c9c961af4b"`;

exports[`Node Wallet Test HDWallet: Node derivePath works when paths provided absolutely 1`] = `"148c7da67c48a8c6da464ba8b3619c2fd4037a43"`;

exports[`Node Wallet Test HDWallet: Node derivePath works when paths provided absolutely 2`] = `"148c7da67c48a8c6da464ba8b3619c2fd4037a43"`;

exports[`Node Wallet Test HDWallet: Node derivePath works when paths provided incrementally 1`] = `"3654cafa0b49eef056adc7dce5f745b08f98498d"`;

exports[`Node Wallet Test HDWallet: Node derivePath works when paths provided incrementally 2`] = `"3654cafa0b49eef056adc7dce5f745b08f98498d"`;
